In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv(r"D:\Desktop\Data_Science\Pandas\Assignments\Assignment_1\data.csv")

In [3]:
df

,Date,Strategy,Campaign,Strategy Type,Seller,Impressions,Clicks,Spend,Sales,Quotes,Unnamed: 10
0,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,PubMatic,10,0,$0.02,0,0,NaN
1,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Google AdExchange,6859,0,$7.50,0,0,NaN
2,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Microsoft Advertising Exchange,74,0,$0.10,0,0,NaN
3,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Technorati Media,7,0,$0.01,0,0,NaN
4,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Ybrant Digital,5,0,$0.01,0,0,NaN
...,...,...,...,...,...,...,...,...,...,...,...
19114,30-6-2015,Time Of Day,Prospecting,Reach,Google AdExchange,962,0,$1.21,0,0,NaN
19115,30-6-2015,Time Of Day,Prospecting,Reach,OpenX,697,1,$0.64,0,0,NaN
19116,30-6-2015,Time Of Day,Prospecting,Reach,Rubicon,2112,0,$3.02,0,0,NaN
19117,30-6-2015,Time Of Day,Prospecting,Reach,OnScroll Limited,56,0,$0.11,0,0,NaN


In [4]:
df.columns

Index(['Date', 'Strategy', 'Campaign', 'Strategy Type', 'Seller',
       'Impressions', 'Clicks', ' Spend', 'Sales', 'Quotes', 'Unnamed: 10'],
      dtype='object')

In [7]:
df.drop('Unnamed: 10',axis=1,inplace=True)

In [8]:
df

,Date,Strategy,Campaign,Strategy Type,Seller,Impressions,Clicks,Spend,Sales,Quotes
0,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,PubMatic,10,0,$0.02,0,0
1,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Google AdExchange,6859,0,$7.50,0,0
2,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Microsoft Advertising Exchange,74,0,$0.10,0,0
3,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Technorati Media,7,0,$0.01,0,0
4,1-6-2015,Audience - Cautious Homeowners,Prospecting,Audience,Ybrant Digital,5,0,$0.01,0,0
...,...,...,...,...,...,...,...,...,...,...
19114,30-6-2015,Time Of Day,Prospecting,Reach,Google AdExchange,962,0,$1.21,0,0
19115,30-6-2015,Time Of Day,Prospecting,Reach,OpenX,697,1,$0.64,0,0
19116,30-6-2015,Time Of Day,Prospecting,Reach,Rubicon,2112,0,$3.02,0,0
19117,30-6-2015,Time Of Day,Prospecting,Reach,OnScroll Limited,56,0,$0.11,0,0


In [62]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19119 entries, 0 to 19118
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date           19119 non-null  datetime64[ns]
 1   Strategy       19119 non-null  object        
 2   Campaign       19119 non-null  object        
 3   Strategy Type  19119 non-null  object        
 4   Seller         19119 non-null  object        
 5   Impressions    19119 non-null  int64         
 6   Clicks         19119 non-null  int64         
 7    Spend         19119 non-null  float64       
 8   Sales          19119 non-null  int64         
 9   Quotes         19119 non-null  int64         
dtypes: datetime64[ns](1), float64(1), int64(4), object(4)
memory usage: 1.5+ MB


In [9]:
df['Date'] = pd.to_datetime(df['Date'], format='%d-%m-%Y')

In [40]:
df[' Spend'] = pd.to_numeric(df[' Spend'].str.replace('$', '').str.strip())

In [11]:
df.dtypes

Date             datetime64[ns]
Strategy                 object
Campaign                 object
Strategy Type            object
Seller                   object
Impressions               int64
Clicks                    int64
 Spend                   object
Sales                     int64
Quotes                    int64
dtype: object

In [14]:
type(df)

pandas.core.frame.DataFrame

In [15]:
df['Strategy'].unique()

array(['Audience - Cautious Homeowners', 'Audience - Demo - Aged 18-24',
       'Audience - In Market - Insurance',
       'Audience - in-Market - Holidays', 'Audience - Mums',
       'Audience Overlap - Car Brand - Ford',
       'Audience Overlap - Festival Lovers',
       'Audience Overlap - Home Movers', 'Auto - Keywords',
       'Brand Keywords', 'Broad Reach', 'eBay', 'Gumtree',
       'Homepage 1 Day Recency', 'Homepage 1 Hour Recency',
       'Homepage 10 Minute Recency', 'Homepage 1-2 Day Recency',
       'Homepage 14 Day Recency', 'Homepage 2-13 Day Recency',
       'Homepage Audience Overlay', 'Homepage Performing Domains',
       'Homepage Unreached Users', 'Insurance - Keywords',
       'Insurance Keywords Audience', 'MoneySupermarket',
       'Overlap Domains', 'Performing Domains', 'Performing Geos',
       'Performing Keywords', 'Thisismoney', 'Time Of Day'], dtype=object)

In [20]:
dir(pd.DataFrame)

['T',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__arrow_c_stream__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__dataframe__',
 '__dataframe_consortium_standard__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__firstlineno__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',
 '__pandas_priority__',
 '__pos__

In [21]:
df['Strategy'].value_counts()

Strategy
Homepage 2-13 Day Recency              1933
Homepage 14 Day Recency                1918
Homepage Unreached Users               1778
Insurance Keywords Audience            1605
Gumtree                                1587
Audience - In Market - Insurance       1326
Homepage 1 Day Recency                 1236
Audience - in-Market - Holidays        1215
Homepage 1-2 Day Recency               1028
Audience - Demo - Aged 18-24            700
Audience Overlap - Home Movers          617
Homepage 1 Hour Recency                 522
Time Of Day                             517
Audience Overlap - Festival Lovers      512
MoneySupermarket                        389
Audience Overlap - Car Brand - Ford     366
Audience - Mums                         304
Broad Reach                             247
Homepage Performing Domains             246
Performing Geos                         229
Performing Keywords                     191
Overlap Domains                         163
Auto - Keywords        

In [22]:
df['Strategy Type'].unique()

array(['Audience', 'Context', 'Reach', 'Retargeting'], dtype=object)

In [23]:
df['Strategy Type'].value_counts()

Strategy Type
Retargeting    8784
Audience       6728
Context        2614
Reach           993
Name: count, dtype: int64

In [24]:
df['Seller'].unique()

array(['PubMatic', 'Google AdExchange', 'Microsoft Advertising Exchange',
       'Technorati Media', 'Ybrant Digital', 'OpenX', 'Rubicon',
       'Unanimis Consulting Limited', 'TLV Media Online Ltd', 'AdExtent',
       'Yellowhammer Media Group Inc.',
       'Global Digital Markets (Ve Interactive)', 'Admetrics Media Ltd',
       'OnScroll Limited', 'Switch Concepts Limited', 'AOL Marketplace',
       'PulsePoint, Inc.', 'Improve Digital', 'bRealTime (select)',
       'Sovrn', 'Adtegrity', 'Lifestreet', 'Criteo',
       'The Publisher Trading Desk Ltd.', 'Admeta AB', 'Defy Media',
       'District M', 'Protein UK Limited', 'Dakine Media LLC',
       'Green Light Media, Inc', 'Programmatic s.r.o.',
       '24/7 Media France', 'MonPar B.V.', 'Brightcom Media',
       'AD2ONE LTD', 'Metric Science', 'The Blogger Network',
       'Yamaka Media Group Ltd', 'Kyston', 'Auto Trader Limited',
       'Adshoard Premium', '33Across Real Display', 'CrimTan',
       'UnderDog Media', 'Bandsintown A

In [25]:
df['Seller'].nunique()

162

In [28]:
df['Seller'].value_counts().sort_values(ascending=False).head(20)

Seller
Google AdExchange                 667
Rubicon                           641
OpenX                             623
OnScroll Limited                  489
Auto Trader Limited               480
Microsoft Advertising Exchange    479
AOL Marketplace                   440
AD2ONE LTD                        402
Switch Concepts Limited           387
Improve Digital                   365
Yellowhammer Media Group Inc.     356
Technorati Media                  354
TLV Media Online Ltd              332
District M                        329
Metric Science                    311
Sovrn                             307
Unanimis Consulting Limited       271
Ybrant Digital                    270
AdExtent                          269
Vibrant Media Inc.                264
Name: count, dtype: int64

In [29]:
df['Seller'].value_counts().sort_values(ascending=False).tail(20)

Seller
Branovate Ltd                                            4
Adtomatik                                                3
Webisaba Select                                          3
Fidelity Media                                           3
Altitude Digital Select                                  3
Ad.net, Inc.                                             2
playbuzz Ltd.                                            2
Tightrope Interactive                                    2
Adhood                                                   2
The Globe and Mail Inc.                                  2
Merriam-Webster, Incorporated, a Delaware Corporation    1
ReachJunction Media Limited                              1
MyLikes Inc.                                             1
Clique Media FZ LLC                                      1
Specific Media                                           1
Yellow Pages Group                                       1
Local Corporation                                

In [35]:
df[['Seller',' Spend']].sort_values(by=' Spend',ascending=False)

,Seller,Spend
5277,AOL Marketplace,$91.52
17470,Google AdExchange,$9.98
17527,Google AdExchange,$9.97
10286,Google AdExchange,$9.96
12591,Google AdExchange,$9.95
...,...,...
14349,Ambient Digital Pte Limited,$0.00
6233,Orbit Interactive,$0.00
6232,Admeta AB,$0.00
6230,Adtegrity,$0.00


In [42]:
top_seller_by_spend=df.groupby('Seller').agg({' Spend':'sum'})

In [46]:
top_seller_by_spend.sort_values(by=' Spend',ascending=False).head(20)

,Spend
Seller,
Google AdExchange,6214.83
Rubicon,4789.00
OpenX,4148.00
AOL Marketplace,1808.67
Facebook,854.64
Microsoft Advertising Exchange,675.39
OnScroll Limited,352.71
Auto Trader Limited,281.72
Switch Concepts Limited,217.75


In [48]:
df1 = df.groupby('Seller')[[' Spend', 'Sales', 'Impressions', 'Clicks', 'Quotes']].sum()

In [50]:
df1.sort_values(by=[' Spend','Sales','Impressions','Clicks','Quotes'],ascending=False).head(10)

,Spend,Sales,Impressions,Clicks,Quotes
Seller,,,,,
Google AdExchange,6214.83,9,2563336,815,46
Rubicon,4789.00,1,1687088,507,19
OpenX,4148.00,7,2419543,589,27
AOL Marketplace,1808.67,1,1215344,616,19
Facebook,854.64,0,262014,199,2
Microsoft Advertising Exchange,675.39,6,113751,101,38
OnScroll Limited,352.71,2,180988,165,26
Auto Trader Limited,281.72,7,111992,50,24
Switch Concepts Limited,217.75,1,112625,57,14


In [51]:
df2 = df.groupby('Strategy')[[' Spend', 'Sales', 'Impressions', 'Clicks', 'Quotes']].sum()
df2.sort_values(by=[' Spend','Sales','Impressions','Clicks','Quotes'],ascending=False).head(10)

,Spend,Sales,Impressions,Clicks,Quotes
Strategy,,,,,
Gumtree,4017.35,14,3140256,580,91
Broad Reach,3067.63,4,2424892,913,9
Homepage 2-13 Day Recency,2533.92,9,423827,315,77
Homepage 14 Day Recency,2393.03,17,670979,330,115
Homepage 1 Day Recency,1380.12,8,155666,156,60
Homepage Unreached Users,1373.22,28,462925,215,139
Homepage Audience Overlay,854.64,0,262014,199,2
Homepage 1-2 Day Recency,731.10,10,89712,82,47
Insurance Keywords Audience,622.15,1,259108,65,66


In [52]:
df2 = df.groupby('Strategy Type')[[' Spend', 'Sales', 'Impressions', 'Clicks', 'Quotes']].sum()
df2.sort_values(by=[' Spend','Sales','Impressions','Clicks','Quotes'],ascending=False).head(10)

,Spend,Sales,Impressions,Clicks,Quotes
Strategy Type,,,,,
Retargeting,9997.64,79,2144499,1379,495
Context,4925.29,20,3315539,678,129
Reach,3438.12,10,2823140,1306,56
Audience,1631.25,15,847800,215,394


In [59]:
df.groupby(['Strategy','Strategy Type']).get_group(('Time Of Day', 'Reach'))

,Date,Strategy,Campaign,Strategy Type,Seller,Impressions,Clicks,Spend,Sales,Quotes
18602,2015-06-01,Time Of Day,Prospecting,Reach,PubMatic,2952,1,2.63,0,0
18603,2015-06-01,Time Of Day,Prospecting,Reach,Technorati Media,1097,1,0.75,0,0
18604,2015-06-01,Time Of Day,Prospecting,Reach,Yellowhammer Media Group Inc.,1215,3,1.11,0,0
18605,2015-06-01,Time Of Day,Prospecting,Reach,Switch Concepts Limited,4694,4,3.05,0,0
18606,2015-06-01,Time Of Day,Prospecting,Reach,AOL Marketplace,4837,3,4.72,0,0
...,...,...,...,...,...,...,...,...,...,...
19114,2015-06-30,Time Of Day,Prospecting,Reach,Google AdExchange,962,0,1.21,0,0
19115,2015-06-30,Time Of Day,Prospecting,Reach,OpenX,697,1,0.64,0,0
19116,2015-06-30,Time Of Day,Prospecting,Reach,Rubicon,2112,0,3.02,0,0
19117,2015-06-30,Time Of Day,Prospecting,Reach,OnScroll Limited,56,0,0.11,0,0


In [60]:
df.groupby(['Strategy','Strategy Type']).size()

Strategy                             Strategy Type
Audience - Cautious Homeowners       Audience           83
Audience - Demo - Aged 18-24         Audience          700
Audience - In Market - Insurance     Audience         1326
Audience - Mums                      Audience          304
Audience - in-Market - Holidays      Audience         1215
Audience Overlap - Car Brand - Ford  Audience          366
Audience Overlap - Festival Lovers   Audience          512
Audience Overlap - Home Movers       Audience          617
Auto - Keywords                      Context           135
Brand Keywords                       Context             3
Broad Reach                          Reach             247
Gumtree                              Context          1587
Homepage 1 Day Recency               Retargeting      1236
Homepage 1 Hour Recency              Retargeting       522
Homepage 1-2 Day Recency             Retargeting      1028
Homepage 10 Minute Recency           Retargeting        94
Homep

In [61]:
df2.index

Index(['Audience', 'Context', 'Reach', 'Retargeting'], dtype='object', name='Strategy Type')

In [67]:
d1=df.groupby('Campaign')[["Impressions","Clicks","Quotes","Sales"," Spend"]].sum()

In [68]:
d1

,Impressions,Clicks,Quotes,Sales,Spend
Campaign,,,,,
Prospecting,6986479,2199,579,45,9994.66
Retargeting,2144499,1379,495,79,9997.64


In [70]:
d1['CPA']=d1[' Spend']/d1['Sales']
d1['CPM']=(d1[' Spend']/d1['Impressions'])*1000



## 🧠 Part 1 — Explaining the Task Using the **STAR Framework**

> 🎯 Goal: Demonstrate analytical thinking + clear communication

---

### **S — Situation**

> “I was given a dataset containing performance data for multiple ad campaigns, including metrics like impressions, clicks, quotes, spend, and sales.
> The goal was to evaluate how different campaign types — *Prospecting* and *Retargeting* — performed overall.”

💬 (This sets the business context clearly — that you’re analyzing campaign efficiency.)

---

### **T — Task**

> “My task was to aggregate the data by campaign and compute key performance metrics like **CPA** (Cost per Acquisition) and **CPM** (Cost per 1,000 impressions), so that I could compare efficiency between Prospecting and Retargeting campaigns.”

💬 (Here, you highlight your technical + analytical objective — summarize and compare.)

---

### **A — Action**

> “Using Pandas, I grouped the dataset by the *Campaign* column and summed up numerical fields such as Impressions, Clicks, Quotes, Spend, and Sales.
> Then, I derived:
>
> * **CPA = Spend / Sales** to measure cost-efficiency of generating a sale, and
> * **CPM = (Spend / Impressions) × 1000** to measure cost of reaching 1,000 users.
>
> This gave me an aggregated view of campaign performance across all strategies.”

💬 (This demonstrates both your technical fluency and the reasoning behind each calculation.)

---

### **R — Result**

> “The aggregation revealed that **Retargeting** campaigns achieved a much lower CPA ($126.55) compared to **Prospecting** ($222.10),
> even though Retargeting had fewer impressions.
>
> This means Retargeting was significantly more cost-efficient in driving conversions, while Prospecting had a lower CPM ($1.43) — indicating it was better for broad audience reach.
>
> Based on this, I would recommend allocating a higher portion of the budget toward Retargeting campaigns for conversion efficiency, while maintaining some spend on Prospecting for awareness.”

💬 (You’ve connected the numbers → insights → business decision → actionable recommendation.)

---

✅ **Final 30-second STAR summary for the interview:**

> “I grouped performance data by Campaign to evaluate Prospecting vs Retargeting.
> After aggregating metrics like Impressions, Clicks, Spend, and Sales, I calculated CPA and CPM to assess efficiency.
> The results showed that Retargeting delivered sales at nearly half the cost of Prospecting (CPA $126 vs $222), though Prospecting achieved cheaper reach (lower CPM).
> Hence, I’d recommend shifting more budget toward Retargeting for conversion efficiency while maintaining limited Prospecting for top-funnel awareness.”

---

## 📊 Part 2 — Insights From the Aggregated Result

| **Campaign**    | **Impressions** | **Clicks** | **Quotes** | **Sales** | **Spend ($)** | **CPA ($)** | **CPM ($)** |
| :-------------- | --------------: | ---------: | ---------: | --------: | ------------: | ----------: | ----------: |
| **Prospecting** |       6,986,479 |      2,199 |        579 |        45 |      9,994.66 |  **222.10** |    **1.43** |
| **Retargeting** |       2,144,499 |      1,379 |        495 |        79 |      9,997.64 |  **126.55** |    **4.66** |

---

### 🔹 **1. Funnel Interpretation**

| Funnel Stage          | Key Metric  | Prospecting | Retargeting | Insight                                                                |
| :-------------------- | :---------- | ----------: | ----------: | :--------------------------------------------------------------------- |
| **Awareness**         | Impressions |       6.98M |       2.14M | Prospecting reached 3× more users → good for awareness                 |
| **Interest (Clicks)** | Clicks      |       2,199 |       1,379 | Engagement levels somewhat similar considering spend                   |
| **Intent (Quotes)**   | Quotes      |         579 |         495 | Despite fewer impressions, Retargeting generated nearly as many quotes |
| **Conversion**        | Sales       |          45 |          79 | Retargeting converted 1.7× more sales with fewer impressions           |

✅ Retargeting clearly captures higher intent users and drives better conversions.

---

### 🔹 **2. Efficiency Metrics**

* **CPA (Cost per Sale)**

  * Prospecting: $222 → High cost per conversion
  * Retargeting: $126 → 43% cheaper

* **CPM (Cost per 1000 Impressions)**

  * Prospecting: $1.43 → Cheap reach (broad audience)
  * Retargeting: $4.66 → More expensive reach, but higher ROI

✅ So, Prospecting = good *reach*, Retargeting = good *returns*.

---

### 🔹 **3. Budget Allocation Insight**

If total future budget = $20,000, you can allocate **inversely proportional to CPA:**

| Campaign    |    CPA | Weight (1/CPA) | % of Budget | New Budget |
| :---------- | -----: | -------------: | ----------: | ---------: |
| Prospecting | 222.10 |         0.0045 |         36% |     $7,200 |
| Retargeting | 126.55 |         0.0079 |         64% |    $12,800 |

✅ Recommendation:

* Increase Retargeting budget from 50% → 65% (conversion efficiency).
* Keep Prospecting at ~35% to continue filling top of the funnel.

---

### 🔹 **4. Business-Level Interpretation**

| Perspective           | Insight                                                               |
| :-------------------- | :-------------------------------------------------------------------- |
| **Efficiency**        | Retargeting converts better and cheaper (lower CPA).                  |
| **Scale**             | Prospecting has wider reach but weak conversion.                      |
| **Optimization Plan** | Reallocate spend — more toward Retargeting, less on Prospecting.      |
| **Next Step**         | Improve creative and audience targeting in Prospecting to reduce CPA. |

---

## 🧩 3️⃣ Optional — How to Present This in the Interview

When they ask *“What did you find?”* → answer in 4 structured sentences:

> “After aggregating the data by Campaign, I compared Prospecting and Retargeting.
> Prospecting had broader reach (7M impressions) but a high CPA of $222, indicating poor efficiency.
> Retargeting, though smaller in scale, achieved lower CPA ($126) and 1.7× more conversions.
> Hence, I’d recommend shifting 15–20% of the Prospecting budget to Retargeting to improve ROI while maintaining awareness coverage.”

---

✅ **Key Takeaway for Interview**

* STAR = clear thinking and structure.
* Insights = data-driven + actionable.
* You show understanding of **funnel dynamics, performance metrics, and budget trade-offs.**

---




In [71]:
d1

,Impressions,Clicks,Quotes,Sales,Spend,CPA,CPM
Campaign,,,,,,,
Prospecting,6986479,2199,579,45,9994.66,222.103556,1.430572
Retargeting,2144499,1379,495,79,9997.64,126.552405,4.661993




Here’s your dataset again for reference:

| **Campaign**    | **Impressions** | **Clicks** | **Quotes** | **Sales** | **Spend ($)** | **CPA ($)** | **CPM ($)** |
| :-------------- | --------------: | ---------: | ---------: | --------: | ------------: | ----------: | ----------: |
| **Prospecting** |       6,986,479 |      2,199 |        579 |        45 |      9,994.66 |  **222.10** |    **1.43** |
| **Retargeting** |       2,144,499 |      1,379 |        495 |        79 |      9,997.64 |  **126.55** |    **4.66** |

---

# 🧭 1️⃣ Reach % Comparison (Impressions)

**Reach** measures *how many users were exposed* to the ad.

[
\text{Reach % Difference} = \frac{\text{Prospecting Impr.} - \text{Retargeting Impr.}}{\text{Retargeting Impr.}} \times 100
]

[
= \frac{6{,}986{,}479 - 2{,}144{,}499}{2{,}144{,}499} \times 100 = 225.8%
]

✅ **Insight:**
Prospecting delivered **~226% more impressions** than Retargeting (≈ 3.26× higher reach).

💬 *Interpretation:*

* Prospecting is doing its job — **broad reach** and **top-funnel awareness**.
* But higher reach doesn’t mean efficiency; it’s just audience exposure.
* Good for filling the funnel with new potential customers.

---

# 💰 2️⃣ CPM % Comparison (Cost per 1,000 Impressions)

[
\text{CPM Difference} = \frac{4.66 - 1.43}{1.43} \times 100 = 225.9%
]

✅ **Insight:**
Retargeting’s CPM is **~226% higher** than Prospecting’s.

💬 *Interpretation:*

* Retargeting costs **more per 1000 impressions**, meaning it targets **narrow, premium, high-intent users**.
* That’s expected — retargeting audiences are smaller but more valuable.
* High CPM is **not bad** if it leads to better conversion efficiency (which it does, as shown by CPA).

---

# ⚖️ 3️⃣ CPA % Comparison (Cost per Acquisition)

[
\text{CPA Difference} = \frac{222.10 - 126.55}{126.55} \times 100 = 75.5%
]

✅ **Insight:**
Prospecting’s CPA is **~75.5% higher** than Retargeting’s.
OR equivalently, Retargeting’s CPA is **43% lower** than Prospecting’s.

💬 *Interpretation:*

* Retargeting drives **sales nearly twice as efficiently**.
* For every $100 spent, Retargeting gets almost **1.7× more sales**.
* Indicates strong bottom-funnel performance — you’re spending less to convert more.

---

# 📈 4️⃣ Combined Insights Summary

| Metric                  | Prospecting | Retargeting |       % Difference | Key Takeaway                                         |
| :---------------------- | ----------: | ----------: | -----------------: | :--------------------------------------------------- |
| **Reach (Impressions)** |       6.98M |       2.14M |  Prospecting +226% | Broad reach for awareness                            |
| **CPM ($)**             |        1.43 |        4.66 |  Retargeting +226% | Higher cost per impression due to narrower targeting |
| **CPA ($)**             |       222.1 |      126.55 | Prospecting +75.5% | Retargeting 43% cheaper per sale → better efficiency |

---

# 🧠 5️⃣ How to Present These Insights Verbally (Interview Version)

> “When I compared Prospecting and Retargeting campaigns, I found that Prospecting achieved about **3× higher reach** with much cheaper CPM ($1.4 vs $4.6), showing it’s effective for awareness.
> However, Retargeting had a **43% lower CPA**, meaning it converted far more efficiently despite the higher cost per impression.
>
> This shows Prospecting is strong at the top of the funnel, while Retargeting dominates the bottom of the funnel — so I’d reallocate a larger share of future budget to Retargeting for efficiency, while keeping Prospecting limited for reach and audience growth.”

---

✅ **Final Takeaway (for your interview answer):**

| Funnel Stage      | Campaign    | Efficiency                                | Budget Decision               |
| :---------------- | :---------- | :---------------------------------------- | :---------------------------- |
| **Top Funnel**    | Prospecting | Wide reach (+226%) but costly CPA (+75%)  | Maintain smaller share (~30%) |
| **Bottom Funnel** | Retargeting | High CPM (+226%) but efficient CPA (–43%) | Scale budget (~70%)           |

---



# Task-2
- metric: Sales

In [80]:
p_df=df[df['Campaign']=="Prospecting"]
strategy=p_df.groupby('Strategy')['Sales'].sum().sort_values(ascending=False)
strategy.head(3)

Strategy
Gumtree                            14
Audience - in-Market - Holidays     6
Audience - Mums                     6
Name: Sales, dtype: int64

In [81]:
strategy_type=p_df.groupby('Strategy Type')['Sales'].sum().sort_values(ascending=False)
strategy_type.nlargest(3)

Strategy Type
Context     20
Audience    15
Reach       10
Name: Sales, dtype: int64

In [83]:
Seller=p_df.groupby('Seller')['Sales'].sum().sort_values(ascending=False)
Seller.head(3)

Seller
Adtegrity            6
Google AdExchange    6
Improve Digital      3
Name: Sales, dtype: int64

Excellent 🔥 — this “Top 3 Performer” task is your **Task-2**, and it’s designed to check not just your coding ability but also your **analytical reasoning** and **decision-making mindset** — exactly what MiQ looks for in a Trader Intern.

Let’s go step-by-step to craft your **STAR-based explanation**, extract **insights**, and explain **why you used total sales** as the performance measure — along with **other possible methods** and their pros/cons.

---

# 🧠 STAR-Based Explanation for Task-2

---

### **S — Situation**

> “After analyzing overall campaign performance in Task-1, I wanted to identify the **top-performing strategies, sellers, and strategy types** within the Prospecting campaign.
> Prospecting generally targets new audiences for awareness and lead generation, so it’s important to know which elements are actually delivering the most conversions (sales).”

💬 *(Sets context — focusing on Prospecting’s internal efficiency.)*

---

### **T — Task**

> “My task was to find the **top 3 performers** in three dimensions — **Strategy**, **Seller**, and **Strategy Type** — to identify which ones contributed most to overall sales volume.”

💬 *(Defines analytical objective clearly.)*

---

### **A — Action**

> “I filtered the dataset for rows where *Campaign = Prospecting*, grouped the data separately by Strategy, Seller, and Strategy Type, summed up total Sales for each group, and then used the `nlargest(3)` function in Pandas to extract the top 3 in each category.
> This simple aggregation helps quickly identify which entities drive maximum conversions in Prospecting.”

💬 *(Highlights logic, simplicity, and reproducibility.)*

---

### **R — Result**

#### **Results Obtained**

| Category          | Top 3 Performers                                          | Total Sales |
| :---------------- | :-------------------------------------------------------- | :---------: |
| **Strategy**      | Gumtree, Audience – Mums, Audience – in-Market – Holidays |   14, 6, 6  |
| **Seller**        | Adtegrity, Google AdExchange, Improve Digital             |   6, 6, 3   |
| **Strategy Type** | Context, Audience, Reach                                  |  20, 15, 10 |

---

### 🧩 **Insights Derived**

#### 1️⃣ **Top 3 Strategies**

* **Gumtree** outperformed others with **14 sales**, contributing the largest share of conversions among Prospecting strategies.
* Audience-based strategies (“Mums”, “In-Market – Holidays”) also performed well, showing **targeted audience segments** outperform broad ones.

💡 *Insight:* Targeted audiences outperform generic placements even in Prospecting, implying that contextual personalization improves conversion even in top-funnel campaigns.

---

#### 2️⃣ **Top 3 Sellers**

* **Adtegrity** and **Google AdExchange** each drove **6 sales**, suggesting these sellers offer better inventory quality for Prospecting.
* **Improve Digital** followed with **3 sales**, indicating moderate efficiency.

💡 *Insight:* Seller quality directly impacts lead generation. Platforms like Google AdX deliver both **scale and relevance**.

---

#### 3️⃣ **Top 3 Strategy Types**

* **Contextual targeting** led with **20 sales**, followed by **Audience** (15) and **Reach** (10).
* Indicates **contextual and audience-based approaches are stronger** even in early funnel stages.

💡 *Insight:* Data-driven and contextual Prospecting works better than purely reach-based campaigns.

---

✅ **Summary of Results**

> “Contextual and Audience-based strategies drive higher conversions than Reach campaigns.
> Sellers like Adtegrity and Google AdX consistently appear among top performers, and targeted strategies such as Gumtree deliver the highest sales volume.”

---

# 🧩 Why I Chose *Total Sales* as Performance Metric

### **Pros (Advantages)**

| Reason                                   | Explanation                                                                                     |
| :--------------------------------------- | :---------------------------------------------------------------------------------------------- |
| **Direct measure of conversion success** | Sales represent actual business outcomes, not intermediate actions like clicks or quotes.       |
| **Simple and interpretable**             | Easy for non-technical stakeholders to understand and visualize.                                |
| **Ideal for Prospecting comparison**     | Helps identify which early-stage strategies are effectively turning awareness into conversions. |

### **Cons (Limitations)**

| Limitation                             | Explanation                                                          |
| :------------------------------------- | :------------------------------------------------------------------- |
| **Ignores efficiency (CPA)**           | A strategy with more sales might still be less cost-effective.       |
| **Skews toward high-spend strategies** | High-budget campaigns may appear “top-performing” just due to scale. |
| **Doesn’t capture engagement quality** | CTR, CVR, or Quote Rate differences aren’t visible here.             |

---

# ⚙️ Other Methods to Find Top Performers

### **1️⃣ CPA-Based Ranking (Efficiency-Focused)**

[
\text{CPA} = \frac{\text{Spend}}{\text{Sales}}
]
✅ **Pros**

* Reflects *cost efficiency* — lower CPA means better ROI.
* Prioritizes strategies delivering more conversions per dollar.

❌ **Cons**

* Can overrate small-sample strategies with very few sales (unstable CPA).
* Ignores overall scale (sales volume).

---

### **2️⃣ Weighted Efficiency (Sales × 1/CPA)**

[
\text{Weighted Score} = \text{Sales} \times \frac{1}{\text{CPA}}
]
✅ **Pros**

* Balances **scale and efficiency** simultaneously.
* Prevents bias toward large-spend or small-CPA strategies alone.

❌ **Cons**

* Needs both metrics clean and reliable (avoid zero-sales division).
* Slightly more complex to explain to non-analysts.

---

### **3️⃣ Composite Score (Normalized Metrics)**

[
\text{Score} = w_1 \times \text{Sales} + w_2 \times (1/\text{CPA}) + w_3 \times \text{CTR}
]
✅ **Pros**

* Multi-dimensional — considers volume, cost, and engagement.
* Used in advanced campaign optimization models.

❌ **Cons**

* Needs proper weight tuning (can introduce bias).
* Harder to communicate in simple dashboards.

---

### **4️⃣ ROI or ROAS-Based Ranking (Profit-Driven)**

If revenue per sale known:
[
\text{ROAS} = \frac{\text{Revenue}}{\text{Spend}}
]
✅ **Pros**

* True profitability indicator.
* Great for optimizing budget allocation.

❌ **Cons**

* Requires accurate revenue data (often unavailable in task datasets).

---

# 📈 Insights from Each Attribute’s Top Performers

| Attribute         | Insight from Top 3                                          | What It Means for a Trader                                             |
| :---------------- | :---------------------------------------------------------- | :--------------------------------------------------------------------- |
| **Strategy**      | “Gumtree” dominates; audience-based strategies perform well | Indicates targeted placements outperform broad ads even in Prospecting |
| **Seller**        | Google AdX & Adtegrity top the list                         | These platforms provide better traffic quality and conversion          |
| **Strategy Type** | Context & Audience outperform Reach                         | Investing in contextual relevance and data segments yields higher ROI  |

---

# 🧭 Interview-Ready STAR Summary (1-Minute Version)

> “In Task-2, I analyzed Prospecting data to identify top-performing strategies, sellers, and strategy types.
> I grouped by each attribute and used total sales to find the top 3 performers because sales are the most direct indicator of campaign success.
> The results showed that Gumtree, Google AdExchange, and Contextual targeting led in conversions — meaning targeted and contextually relevant campaigns outperform broad reach campaigns in Prospecting.
>
> The advantage of using total sales is its simplicity and alignment with business goals, though it ignores efficiency (CPA).
> Alternative approaches like CPA-based or weighted scoring methods could provide a more balanced view by combining cost efficiency and scale.”

---

✅ **Key Takeaways (for your interview):**

* You can explain **methodology, logic, and trade-offs.**
* You connect **data → insight → business action.**
* You understand **alternative approaches** and can justify **why you chose one.**

---

Would you like me to help you craft the **Task-3 STAR answer** (e.g., when you compared Prospecting vs Retargeting or computed top performers by CPA)?


Perfect ✅ — this is exactly how you should structure your Task-3 explanation in the MiQ interview.
This task demonstrates **your ability to measure efficiency (CPA)**, **handle data edge cases**, and **justify the choice of weighted averages** — all of which are highly valued for Trader Interns.

Below is a complete **STAR-based explanation** + intuitive **example** + reasoning for **weighted CPA and NaN handling** 👇

---

# 🧠 STAR-Based Explanation — *Top Performer by CPA*

---

### **S — Situation**

> “After identifying the top performers by total sales in Task-2, I wanted to analyze which strategies, sellers, and strategy types were the most **cost-efficient** — that is, which ones drove conversions at the lowest average cost per acquisition (CPA).
> Since Prospecting campaigns vary in scale and spend, I needed a fair way to compare CPA performance across uneven datasets.”

💬 *(Shows your analytical reasoning and why you moved beyond total sales.)*

---

### **T — Task**

> “My task was to compute the **top 3 performers by CPA** across three dimensions — *Strategy*, *Strategy Type*, and *Seller*.
> However, rather than a simple average, I used a **weighted CPA** to ensure larger-volume strategies had proportional influence on the results.”

💬 *(Defines the objective precisely and introduces the improved metric.)*

---

### **A — Action**

> “I first removed rows where CPA was `NaN` — those corresponded to zero-sales records, where CPA is undefined.
> Then I grouped the valid Prospecting data by each attribute and calculated a **weighted CPA** using the formula:
>
> [
> \text{Weighted CPA} = \frac{\sum(\text{CPA}_i \times \text{Sales}_i)}{\sum(\text{Sales}_i)}
> ]
>
> This ensures that strategies contributing more sales have higher weight in the final CPA, avoiding distortion by small-sample strategies with very few conversions.
> Finally, I used `.nsmallest(3)` to extract the top 3 most efficient performers with the lowest overall CPA.”

💬 *(Shows both your technical implementation and reasoning behind each step.)*

---

### **R — Result**

> “The output identified the 3 strategies, sellers, and strategy types with the **lowest overall CPA**, meaning they achieved conversions at the least cost.
> These are the most **cost-efficient performers**, ideal candidates for **increased budget allocation or scaling**.”

💬 *(Emphasizes insight and business impact — key for a trading interview.)*

---

# 🧮 1️⃣ How Weighted CPA Works — Simple Example

| Strategy | Sales | Individual CPA ($) | Weighted Impact |
| :------- | ----: | -----------------: | --------------: |
| A        |    10 |                100 |  10×100 = 1,000 |
| B        |   100 |                 50 |  100×50 = 5,000 |
| C        |     5 |                200 |   5×200 = 1,000 |

[
\text{Weighted CPA} = \frac{1{,}000 + 5{,}000 + 1{,}000}{10 + 100 + 5} = \frac{7{,}000}{115} = 60.9
]

✅ So even though Strategy C has a high CPA (200), its low sales (5) don’t skew the overall result.
Weighted CPA = realistic, business-scaled efficiency.

---

# ⚖️ 2️⃣ Why Use Weighted CPA (and Not Simple Average)

| Approach               | What It Does               | Problem                                                                    |
| :--------------------- | :------------------------- | :------------------------------------------------------------------------- |
| **Simple Average CPA** | Takes the mean of all CPAs | Treats all campaigns equally, even if one has 1 sale and another 100 sales |
| **Weighted CPA** ✅     | Weights by number of Sales | Reflects *true cost efficiency* of the total conversions generated         |

💬 *Example Insight:*

> “If one strategy made 2 sales at ₹10 CPA and another made 200 sales at ₹20 CPA, a simple average (₹15) would understate the true cost. Weighted CPA (≈₹19.8) better represents the actual campaign performance.”

---

# 🧩 3️⃣ Handling Edge Cases — When CPA = NaN

**Why CPA becomes NaN:**

* CPA = Spend / Sales
* If `Sales = 0`, division by zero → `NaN`

**How you handled it:**

```python
df_valid = prospecting_df.dropna(subset=["CPA"])
```

✅ **Why handle this:**

* Rows with `NaN` CPA mean **no conversions**, so CPA is *undefined*, not 0.
* Including them would produce false averages and mislead efficiency comparisons.

💬 *Interview Explanation:*

> “I excluded NaN CPA rows since those represent non-converting records.
> They’re valuable for awareness analysis but not for cost-efficiency ranking — including them would artificially inflate average CPA or cause division errors.”

---

# 📊 4️⃣ Insights You’d Present From the Result

* Strategies with the **lowest CPA** are most **cost-efficient** → they drive cheaper conversions.
* Comparing CPA with total sales can show whether a strategy is *efficient but small-scale* or *efficient and scalable*.
* Helps in **budget reallocation** — shifting funds toward high-efficiency, scalable segments.

💬 *Example talking point:*

> “The top CPA performers indicate which strategies provide the highest return on every dollar spent.
> These should receive a larger portion of the Prospecting budget in the next cycle, while high-CPA, low-conversion strategies should be optimized or paused.”

---

# 🧠 5️⃣ STAR Summary (1-Minute Interview Answer)

> “In Task-3, I aimed to find the most cost-efficient performers using CPA as the key metric.
> I filtered out NaN CPA rows caused by zero-sales cases, then grouped by Strategy, Seller, and Strategy Type.
> To ensure fairness, I calculated **weighted CPA** — multiplying each campaign’s CPA by its sales volume, summing across the group, and dividing by total sales.
> This approach gives more weight to larger campaigns and avoids small-sample distortions.
> The result highlighted which segments achieved the lowest CPA, helping prioritize them for future budget allocation.
>
> Weighted CPA is better than a simple average because it represents true cost efficiency across varying campaign sizes, and excluding NaN CPA ensures we only measure actual conversions.”

---

✅ **Key Takeaways (to memorize for interview):**

* Weighted CPA reflects **scale-adjusted efficiency**.
* Excluding NaN avoids **false or infinite CPAs**.
* Lower CPA = **better efficiency**, but cross-check with Sales volume before scaling.

---

Would you like me to help you add the **insight comparison** — e.g., “Top 3 by Sales vs Top 3 by CPA” — and how to discuss that contrast in your interview (it’s a very common follow-up question)?


Perfect 🔥 — now you’ve completed **Task-2 for the Retargeting campaign**, which mirrors what you did for Prospecting.
This comparison is extremely valuable in the MiQ interview because it shows that you not only analyzed each campaign type, but also **contrasted them strategically** — identifying how *upper-funnel (Prospecting)* and *lower-funnel (Retargeting)* tactics differ.

Let’s go step by step:

---

## 🧠 1️⃣ Recap — Prospecting vs Retargeting (Top Performers by Sales)

| Category          | **Prospecting (Top 3)**       | **Sales** | **Retargeting (Top 3)**        | **Sales** |
| :---------------- | :---------------------------- | --------: | :----------------------------- | --------: |
| **Strategy**      | Gumtree                       |        14 | Homepage Unreached Users       |        28 |
|                   | Audience – Mums               |         6 | Homepage 14 Day Recency        |        17 |
|                   | Audience – In-Market Holidays |         6 | Homepage 1-2 Day Recency       |        10 |
| **Seller**        | Adtegrity                     |         6 | Auto Trader Limited            |         6 |
|                   | Google AdExchange             |         6 | Microsoft Advertising Exchange |         6 |
|                   | Improve Digital               |         3 | AD2ONE LTD                     |         4 |
| **Strategy Type** | Context                       |        20 | Retargeting                    |        79 |
|                   | Audience                      |        15 | —                              |         — |
|                   | Reach                         |        10 | —                              |         — |

---

## ⭐ STAR-Based Explanation

### **S — Situation**

> “After analyzing Prospecting campaign performance, I replicated the same process for the Retargeting campaign to identify which strategies, sellers, and strategy types generated the highest sales.
> Retargeting campaigns target previously engaged users, so their goal is conversion efficiency rather than broad reach.”

---

### **T — Task**

> “My task was to aggregate sales across Retargeting strategies, sellers, and strategy types, identify the top 3 performers by total conversions, and then compare their performance pattern with the Prospecting campaign.”

---

### **A — Action**

> “I filtered the dataset for rows where *Campaign = ‘Retargeting’*, grouped by Strategy, Seller, and Strategy Type, summed the Sales, and used `.nlargest(3)` to extract the top 3 performers for each.
> Then I compared these results to the Prospecting campaign to highlight how upper-funnel and lower-funnel strategies differ.”

---

### **R — Result**

> “The top-performing Retargeting strategies — *Homepage Unreached Users*, *Homepage 14 Day Recency*, and *Homepage 1-2 Day Recency* — together delivered significantly higher total sales (28 + 17 + 10 = 55) than Prospecting’s top strategies (14 + 6 + 6 = 26).
> This confirms that Retargeting is far more conversion-efficient.”

---

## 📊 2️⃣ Insights — Prospecting vs Retargeting Comparison

| Dimension                   | **Prospecting**                           | **Retargeting**                                 | **Comparison & Interpretation**                          |
| :-------------------------- | :---------------------------------------- | :---------------------------------------------- | :------------------------------------------------------- |
| **Objective**               | Acquire new users (awareness)             | Convert interested users (action)               | Complementary funnel stages                              |
| **Top Strategy Theme**      | Audience / Contextual                     | Homepage Recency (1–14 days)                    | Retargeting focuses on *recent visitors*, strong intent  |
| **Top Seller Pattern**      | Open ad exchanges (Google AdX, Adtegrity) | Premium direct sellers (Microsoft, Auto Trader) | Retargeting relies on *high-quality, targeted inventory* |
| **Sales Volume**            | Top 3 = 26 sales                          | Top 3 = 55 sales                                | Retargeting ≈ 2.1× higher conversions                    |
| **Strategy Type Mix**       | Context (20), Audience (15), Reach (10)   | Retargeting (79)                                | Retargeting dominates lower funnel                       |
| **CPA Trend (from Task 1)** | ~$222 per sale                            | ~$126 per sale                                  | Retargeting ≈ 43 % more cost-efficient                   |
| **Budget Focus**            | Lower cost → broader reach                | Higher CPM → better ROI                         | Allocate larger share to Retargeting                     |

---

## 💡 3️⃣ Key Business Insights

1. **Recency Matters** – The best-performing Retargeting strategies (1-day → 14-day recency) show that **users who interacted recently are far more likely to convert**.
   → Budget priority: recent-visitor pools.

2. **Unreached Users Retargeting** – “Homepage Unreached Users” (28 sales) performed best, suggesting that re-engaging high-intent visitors who missed earlier messaging is extremely valuable.

3. **Prospecting Feeds Retargeting** – Prospecting builds initial traffic; Retargeting harvests conversions from that audience.
   → The two are **funnel complements**, not substitutes.

4. **Seller Insights** – Premium networks (Microsoft, Auto Trader) work better for Retargeting due to more controlled placements and higher user relevance, while open exchanges (AdX, Adtegrity) scale Prospecting effectively.

5. **Budget Strategy** –

   * Keep **~30–35 %** budget for Prospecting to maintain audience inflow.
   * Shift **~65–70 %** to Retargeting for conversion efficiency.

---

## ⚖️ 4️⃣ Interview-Ready 1-Minute Answer

> “I replicated my Prospecting analysis for the Retargeting campaign and found that retargeting strategies like *Homepage Unreached Users* and *14-Day Recency* drove nearly double the sales of Prospecting’s top strategies.
> While Prospecting used contextual and audience targeting for reach, Retargeting focused on recent visitors and achieved far higher conversion efficiency.
> Sellers such as Microsoft Ad Exchange and Auto Trader also stood out, showing that premium inventory yields better ROI in lower-funnel campaigns.
> This contrast highlights that Prospecting builds reach and audience pools, whereas Retargeting efficiently converts them — so budgets should be weighted toward Retargeting while keeping Prospecting for funnel inflow.”

---

✅ **Summary Table for Your Slide / Interview Notes**

| Funnel          | Strength                 | Weakness                 | Role in Trading   |
| :-------------- | :----------------------- | :----------------------- | :---------------- |
| **Prospecting** | Wide reach, cheap CPM    | High CPA, low conversion | Build audience    |
| **Retargeting** | High conversion, low CPA | Limited scale            | Monetize audience |

---

Would you like me to prepare a **Task-2 & Task-3 combined STAR summary** (Prospecting Sales vs Retargeting Sales vs CPA) so you can deliver a single coherent 2-minute story in your interview?


Perfect 👏 — this is *exactly* the kind of question MiQ or any ad-tech/data analytics interviewer would ask you to explain conceptually, not just code-wise. Let’s break this into a clear, polished **STAR-style + Insight-based explanation** that you can directly use in your interview.

---

## 🟩 **S – Situation**

MiQ provided me with campaign performance data containing metrics like *Spend, Quotes, Sales, Strategy,* and *Campaign Type*.
My task was to **allocate a new budget of $15,000** for the *Prospecting* campaign to achieve the **target CPA of $150 per Sale** and **$14 per Quote**.

---

## 🟩 **T – Task**

I had to determine how to **distribute this budget across multiple strategies** under the Prospecting campaign in such a way that:

1. The total stays within $15,000.
2. Strategies with better historical performance (i.e., lower CPA) receive higher allocations.
3. I could project future Quotes and Sales achievable under the new budget.

---

## 🟩 **A – Action**

### 1️⃣ **Filtered Relevant Campaign Data**

I first filtered the dataset to include only the *Prospecting* campaign.

```python
p_df = df2[df2["Campaign"] == "Prospecting"]
```

### 2️⃣ **Calculated Historical CPA for Each Strategy**

To measure efficiency:

* `CPA_Quote = Spend / Quotes`
* `CPA_Sale = Spend / Sales`
  I handled edge cases (like division by zero) using `np.where()` to replace invalid divisions with NaN.

### 3️⃣ **Computed Weighted Average CPA per Strategy**

Instead of a simple mean CPA, I used a **weighted CPA**, giving higher importance to strategies that delivered more conversions.
This avoids bias from small-volume strategies that had artificially low CPAs.

```python
weighted_CPA = (CPA * conversions).sum() / conversions.sum()
```

### 4️⃣ **Allocated Budget Proportionally (Programmatic Approach)**

To reward efficiency:

* Calculated **weights inversely proportional to CPA** → lower CPA = higher efficiency = more budget.
* Normalized weights so the sum equals the total budget ($15,000).

```python
weight = 1 / strategy_stats["Overall_CPA_Sale"]
strategy_stats["Budget_Allocated"] = (weight / weight.sum()) * total_budget
```

### 5️⃣ **Projected Future Performance**

Based on target CPAs, projected how many Quotes and Sales each strategy could generate:

```python
Projected_Quotes = Budget_Allocated / target_cpa_quote
Projected_Sales = Budget_Allocated / target_cpa_sale
```

---

## 🟩 **R – Result**

| Strategy                         | Budget_Allocated | Projected_Quotes | Projected_Sales |
| -------------------------------- | ---------------- | ---------------- | --------------- |
| Audience - In Market - Insurance | $9,410.91        | 672              | 63              |
| Audience - Mums                  | $4,705.46        | 336              | 31              |
| Time of Day                      | $427.77          | 31               | 3               |
| Performing Geos                  | $376.44          | 27               | 3               |
| Audience - In Market - Holidays  | $35.87           | 3                | 0               |
| Broad Reach                      | $30.75           | 2                | 0               |
| Gumtree                          | $12.80           | 1                | 0               |
| **Total**                        | **$15,000**      | **1072**         | **100**         |

🟢 **Insights:**

* *Audience - In Market - Insurance* and *Audience - Mums* clearly outperform other strategies — they captured **~95%** of the total allocated budget.
* Lower-performing or niche strategies like *Gumtree* and *Broad Reach* received minimal budget, avoiding waste.
* Achieving 100 projected sales aligns with the target CPA of $150.

---

## 💡 **Challenges & How I Solved Them**

| Challenge                                 | Solution                                                                                     |
| ----------------------------------------- | -------------------------------------------------------------------------------------------- |
| **Division by Zero (no sales/quotes)**    | Used `np.where()` to replace divisions by zero with `NaN` and filled missing weights with 0. |
| **Small sample strategies skewing CPA**   | Used **weighted CPA** instead of simple average to ensure fair comparison.                   |
| **Ensuring budget sum = $15,000 exactly** | Normalized weights to make the proportional allocation add up to the total.                  |

---

## 💭 **Why Weighted CPA?**

Weighted CPA gives a more reliable measure of performance because it considers both efficiency and volume.
Example:

* Strategy A: CPA = $50 (2 sales)
* Strategy B: CPA = $60 (100 sales)
  If you take a simple mean, you’d wrongly think A is better — but weighted CPA reveals that B is consistently better at scale.

---

## 💭 **Why Programmatic (Proportional) Allocation**

**Pros:**

* Objective and data-driven — removes bias.
* Automatically scales budgets with performance efficiency.
* Easy to rerun for new budgets or target CPAs.

**Cons:**

* Relies on past performance — may not capture future shifts.
* Doesn’t account for diminishing returns if budget increases too fast.

---

## 💡 **Alternate Approaches (and Why I Chose This One)**

* **Manual heuristic allocation:** based on intuition or static % splits — simple but less data-driven.
* **Optimization model (ROI-based / regression):** more precise but overkill for small datasets.
* ✅ **Chosen:** Weighted inverse-CPA proportional allocation — simple, explainable, scalable.

---

## 🎯 **Final Takeaway / Insights**

This exercise gave me hands-on understanding of **budget optimization, CPA-based performance measurement, and trade-offs between data-driven and heuristic allocation**.
It also deepened my grasp of the **ad-tech funnel (Impressions → Clicks → Quotes → Sales)** and how **strategic budget shifts** can maximize conversions while maintaining CPA goals.

---

Would you like me to make this into a **1-page written report summary** you can attach or mention during your MiQ interview (as a polished explanation of your assignment)?


Perfect — this is exactly the kind of structured analytical problem MiQ wants to see you explain fluently in your **interview**.
Below is a **complete STAR explanation**, your **approach rationale**, **edge-case handling**, and **insights**, all written in clear, professional language — the way you’d present it in an interview or case discussion.

---

## 🧠 **STAR Answer (Situation – Task – Action – Result)**

### **S – Situation:**

MiQ provided historical performance data for different **retargeting strategies** under a digital ad campaign.
Each strategy had its **spend**, **quotes**, and **sales**.
A **new total budget of ₹20,000** was allocated for the next campaign cycle, and the task was to optimally distribute it across these strategies to maximize outcomes under the following targets:

* 🎯 Target **CPA (Cost per Sale)** = ₹150
* 🎯 Target **CPA (Cost per Quote)** = ₹14

The goal was to make this allocation data-driven and performance-oriented.

---

### **T – Task:**

My task was to:

1. Calculate each strategy’s **historical efficiency** using CPA metrics.
2. Allocate the new budget **proportionally to performance**, giving more weight to efficient strategies.
3. Handle missing or zero conversion data carefully.
4. Forecast **projected Quotes and Sales** based on the new budget and target CPAs.

---

### **A – Action:**

#### **Step 1: Data Preparation & Filtering**

Filtered the dataset for **Retargeting** campaigns only:

```python
r_df = df2[df2["Campaign"] == "Retargeting"]
```

Then calculated **CPA for Quotes** and **CPA for Sales** using conditional division:

```python
r_df["CPA_Quote"] = np.where(r_df["Quotes"] != 0, r_df["Spend"] / r_df["Quotes"], np.nan)
r_df["CPA_Sale"]  = np.where(r_df["Sales"]  != 0, r_df["Spend"] / r_df["Sales"],  np.nan)
```

✅ Prevented divide-by-zero issues
✅ Ensured strategies with no conversions don’t distort results

---

#### **Step 2: Compute Weighted CPAs (Efficiency Measure)**

Instead of a simple average CPA, I used **Weighted CPA** to make the metric conversion-volume aware:

```python
weighted CPA = Σ(CPA × conversions) / Σ(conversions)
```

This gave a fairer view — e.g., a strategy with 100 sales and another with 2 shouldn’t influence the average equally.

**Why Weighted CPA?**

* Accounts for both efficiency *and* volume.
* Reduces outlier distortion (like a small strategy with very high/low CPA).

---

#### **Step 3: Proportional Budget Allocation (Programmatic Approach)**

To allocate ₹20,000 fairly:

* Computed weights **inversely proportional to CPA**
  → *Lower CPA = More efficient = Higher weight*

```python
weight = 1 / strategy_stats_2["Overall_CPA_Sale"]
strategy_stats_2["Budget_Allocated"] = (weight / weight.sum()) * total_budget
```

✅ Automatically gave more budget to cost-efficient strategies.
✅ Ensured all budgets sum exactly to ₹20,000.

---

#### **Step 4: Projected Future Performance**

Based on MiQ’s target CPAs:

```python
target_cpa_quote = 14
target_cpa_sale  = 120
```

Estimated expected conversions under optimized allocation:

```python
Projected_Quotes = Budget_Allocated / target_cpa_quote
Projected_Sales  = Budget_Allocated / target_cpa_sale
```

---

#### **Step 5: Edge Case Handling**

| Issue                            | Example                                      | Fix                                                     |
| -------------------------------- | -------------------------------------------- | ------------------------------------------------------- |
| Zero quotes/sales → Infinite CPA | Strategy with 0 conversions                  | Used `np.nan` and replaced with 0 weight                |
| Missing data                     | Some strategies had incomplete spend records | Used fillna(0) to exclude them from weight calculations |
| Extremely high CPA values        | Skewed averages                              | Weighted average reduced their influence                |

---

### **R – Result:**

| Strategy                    | Budget (₹) | Projected Quotes | Projected Sales |
| --------------------------- | ---------- | ---------------- | --------------- |
| Homepage 10-Min Recency     | 8,200      | 586              | 68              |
| Homepage 1-Hour Recency     | 5,467      | 390              | 46              |
| Homepage 1-Day Recency      | 2,573      | 184              | 21              |
| Homepage Unreached Users    | 1,617      | 115              | 13              |
| Homepage Performing Domains | 1,396      | 100              | 12              |
| Homepage 1–2 Day Recency    | 503        | 36               | 4               |
| Homepage 14-Day Recency     | 194        | 14               | 2               |
| Homepage 2–13 Day Recency   | 50         | 4                | 0               |
| Audience Overlay            | 0          | 0                | 0               |

✅ **Total Budget:** ₹20,000
✅ **Projected Quotes:** 1,429
✅ **Projected Sales:** 166

---

## ⚙️ **Why This Approach (Programmatic Weighted Allocation)?**

### **Pros:**

* Objective & data-driven — removes guesswork.
* Scalable — easily extendable if new strategies are added.
* Favors high-performing, cost-efficient strategies.
* Simple to explain, visualize, and defend in an interview.

### **Cons:**

* Assumes linear relationship (Spend ↑ → Conversions ↑) — which may not always hold.
* Doesn’t model diminishing returns — advanced models like regression or optimization can refine it further.

### **Alternative Approaches (and Why I Chose Not to Use Them Yet):**

| Approach                      | Description                    | Reason Not Chosen                                      |
| ----------------------------- | ------------------------------ | ------------------------------------------------------ |
| Equal distribution            | Split evenly                   | Ignores past performance                               |
| Sales/Quote-based weighting   | Allocate by total sales/quotes | Misses cost efficiency                                 |
| Regression-based optimization | Predict sales vs. spend        | Overkill for available data; requires more data points |
| ROAS-based allocation         | Return-oriented                | Not applicable since revenue not given                 |

Hence, the **inverse CPA-weighted proportional method** struck the right balance between simplicity and insight.

---

## 💡 **Insights from Results**

1. **Recency drives conversions:**
   Shorter recency windows (10-min & 1-hour) dominate performance. These users are highly engaged → allocate majority budget here.

2. **Long-tail inefficiency:**
   Strategies like “2–13 Day Recency” and “Audience Overlay” produced negligible results → can be **paused or capped**.

3. **Balanced diversification:**
   Some mid-range strategies (Unreached Users, Performing Domains) still received modest budgets to test secondary audience segments.

4. **Improved conversion expectation:**
   The proportional reallocation maintains cost efficiency while projecting **1,429 Quotes & 166 Sales** under ₹20,000 — hitting the performance targets efficiently.

---

## 🗣️ **Interview-Ready Summary (1-Min Pitch)**

> “In the budget allocation task, I optimized ₹20,000 across retargeting strategies using a proportional, inverse-CPA-based allocation approach.
> I first computed weighted CPAs for fairness, handled zero or missing conversions safely, and allocated more budget to efficient strategies.
> This resulted in stronger focus on high-recency audiences and projected ~1,400 quotes and 166 sales at target CPAs.
> The task was both challenging and insightful — it deepened my understanding of the ad-tech funnel and data-driven decision-making in campaign optimization.”

---

Would you like me to create a **concise “cheat sheet”** version of this — a 6–8 line summary you can read right before your MiQ interview for instant recall?


Excellent — this is the *final performance evaluation* stage of your MiQ assignment, where you analyze **campaign efficiency after budget reallocation**.
Let’s go step-by-step and structure this into a **STAR + insight-based interview explanation** so you can clearly communicate your reasoning and learnings.

---

## 🧠 **STAR-Based Interview Explanation**

### **S – Situation:**

After completing the **budget reallocation** for both *Prospecting* and *Retargeting* campaigns, the next step was to evaluate the **impact on performance metrics** — specifically **Quotes, Sales, CPA (Cost per Acquisition), and CPM (Cost per 1,000 Impressions)**.
The goal was to check how effectively the new budget was distributed and whether it maintained or improved campaign efficiency.

---

### **T – Task:**

1. Compute key post-allocation metrics:

   * Total projected **Quotes** and **Sales** for both campaigns.
   * New **CPA** (based on projected Sales).
   * Updated **CPM** (using historical impressions and proportional scaling).
2. Compare these values against historical metrics to evaluate improvement or consistency.
3. Explain why CPM remained unchanged after allocation.

---

### **A – Action:**

#### **Step 1: Calculated Aggregated Metrics**

```python
total_quotes_prospecting = budget_table['Projected_Quotes'].sum()
total_sales_prospecting  = budget_table['Projected_Sales'].sum()
total_quotes_retargeting = budget_table_2['Projected_Quotes'].sum()
total_sales_retargeting  = budget_table_2['Projected_Sales'].sum()
```

✅ Combined the projections from all strategies under each campaign.

---

#### **Step 2: Calculated New CPAs (Cost per Acquisition – Sales)**

```python
cpa_prospecting = budget_table['Budget_Allocated'].sum() / total_sales_prospecting
cpa_retargeting = budget_table_2['Budget_Allocated'].sum() / total_sales_retargeting
```

✅ This shows the **average cost to acquire one sale** post-allocation.
Prospecting CPA = **₹150.00**, Retargeting CPA = **₹120.48**, indicating the reallocation is cost-efficient.

---

#### **Step 3: Calculated CPM (Cost per 1,000 Impressions)**

```python
cpm_prospecting = (
    budget_table['Budget_Allocated'].sum()
    / ((agg.loc['Prospecting', 'Impressions'] * budget_table['Budget_Allocated'].sum()) / agg.loc['Prospecting', 'Spend'])
) * 1000
```

**Explanation of logic:**

* The idea is to keep **historical impression efficiency** consistent.
* We estimate the *new impressions* proportionally to how the spend changes relative to historical spend:
  [
  \text{New Impressions} = \text{Historical Impressions} \times \frac{\text{New Spend}}{\text{Historical Spend}}
  ]
* Then calculate:
  [
  \text{CPM} = \frac{\text{New Spend}}{\text{New Impressions}} \times 1000
  ]

✅ This keeps CPM consistent with the campaign’s historical cost efficiency per impression.

---

#### **Step 4: Derived Final Metrics Table**

| **Metric**  | **Prospecting** | **Retargeting** |
| ----------- | --------------- | --------------- |
| **Quotes**  | 1,072           | 1,429           |
| **Sales**   | 100             | 166             |
| **CPA (₹)** | 150.00          | 120.48          |
| **CPM (₹)** | 1.43            | 4.66            |

---

### **R – Result:**

✅ **CPA Improvement:**

* Retargeting campaign achieved a lower CPA (₹120.48) — meaning it’s now acquiring customers more efficiently than before.
* Prospecting maintained its target CPA (₹150), indicating balanced allocation.

✅ **CPM Unchanged:**

* CPM stayed the same as historical data because **impressions scale linearly with spend** in the current proportional model.
* No change in media buying efficiency is expected until we optimize for creatives or bidding.

✅ **Total Volume Growth:**

* Combined projected sales = **266 (100 + 166)**
* A significant increase in total conversions with the same spend efficiency.

---

## ⚙️ **Why CPM Remained Unchanged**

**Reason:**
CPM depends on **spend per impression**, not on conversions.
Since the **same platforms, ad formats, and auction mechanisms** are used, and the allocation didn’t change the relative spend-to-impression ratio, the **CPM remains constant** — the campaign efficiency in *reach* stays the same.

If we wanted CPM to change, we’d need to:

* Switch to more efficient inventory (e.g., cheaper publishers or ad exchanges).
* Optimize creatives or bidding strategies.
* Change targeting granularity.

In this task, only **budget distribution** was optimized — not media cost — so CPM stays equal to historical CPM.

---

## 💡 **Insights from Results**

1. **Retargeting Outperforms Prospecting:**
   Retargeting yields a lower CPA and higher sales volume, confirming that focusing on previously engaged users delivers better ROI.

2. **Balanced Prospecting Efficiency:**
   Prospecting maintained its CPA target — meaning the model didn’t over-allocate or under-allocate to any inefficient strategy.

3. **Stable CPM Values:**
   The unchanged CPM reflects consistent ad delivery costs — proving the reallocation maintained channel efficiency.

4. **Performance Funnel Understanding:**
   Prospecting drives awareness (Quotes), while Retargeting converts warm leads (Sales) — both complement each other for overall growth.

---

## 🗣️ **Interview-Ready Summary (60 seconds)**

> “After reallocating the budget, I calculated post-allocation metrics — Quotes, Sales, CPA, and CPM — for both campaigns.
> Using the weighted projections, Prospecting achieved 1,072 quotes and 100 sales at a ₹150 CPA, while Retargeting achieved 1,429 quotes and 166 sales at a ₹120 CPA.
> I used proportional scaling based on historical impressions to compute CPM, which remained constant because impression delivery costs didn’t change.
> The analysis showed that retargeting performs more efficiently per acquisition, while both campaigns maintained consistent reach efficiency.”

---

Would you like me to also create a **visual table (before vs after allocation)** format for this — the kind you can show in the Excel workbook or PowerPoint if they ask you to “summarize impact visually”?


Here’s a strong **STAR-based answer** with **insights** for **7a**:

---

### **Question 7a:**

*What type of product do you think the data relates to and why?*

---

### **S – Situation:**

I was provided with a campaign dataset that included metrics such as **Quotes, Sales, Strategy Names, and Budget allocations** for both **Prospecting** and **Retargeting** strategies. The task was to infer what type of product or service was being promoted based on the structure and terminology of the data.

---

### **T – Task:**

My objective was to identify the **nature of the product** behind the campaign by analyzing the **data patterns, strategy naming conventions, and funnel stages** reflected in the dataset.

---

### **A – Action:**

* I observed that the dataset included both **“Quotes”** and **“Sales”** as performance metrics. This suggested a **two-step conversion process**, where customers first request a quote before completing a purchase — a pattern commonly seen in **insurance, real estate, or financial service** products.
* I further analyzed the **strategy names**, which included terms like **“Audience - Cautious Homeowners”**, **“Audience - Mums”**, and **“Insurance - Keyword.”**
* These audience and keyword labels indicated **targeted segments** often associated with **insurance or protection-based financial campaigns**, where different demographics (homeowners, parents, etc.) are targeted with tailored messaging.

---

### **R – Result:**

Based on these observations, I concluded that the campaign most likely relates to an **insurance product or financial service** (e.g., home insurance, life insurance, or family protection plans).

---

### **Insight:**

This exercise highlighted how **campaign structure and funnel metrics** can reveal the nature of a business model even without explicit labels.

* The presence of “Quotes → Sales” conversion stages indicates **high-consideration products** that require **customer engagement before purchase**.
* Audience-based strategies show a **personalized targeting approach**, which is key in **ad-tech and insurance marketing** to optimize reach and conversion efficiency.

---

Would you like me to write similar **STAR + Insight** style answers for the next sub-questions (7b, 7c, 7d)?


Here’s a **STAR-based answer** with clear **insight, explanations, and layman-friendly reasoning** for **Question 7b** 👇

---

### **Question 7b:**

*Suggest at least 3 new possible display advertising strategies for this particular campaign.*

---

### **S – Situation:**

From the previous analysis, the campaign data suggested that the product being promoted was most likely an **insurance or financial service**. The existing strategies in the dataset were focused on **broad audience segments** like *“Cautious Homeowners”* and *“Mums”*, and **keyword-based targeting** such as *“Insurance – Keyword.”*
My goal was to suggest **new display advertising strategies** that could further optimize reach and improve conversion rates.

---

### **T – Task:**

To design **3–4 new display advertising strategies** that align with the nature of the product (insurance/financial) and help in improving engagement, conversion, and cost efficiency.

---

### **A – Action:**

I proposed the following new **display advertising strategies** —

#### **1. Localised Targeting with Tailored Offers**

**Meaning:** Show different ad creatives and offers depending on where a person lives.
**Why:** Insurance and financial products are often influenced by **regional risks** (e.g., flood-prone areas, city traffic, or cost of living).
**Example:** Showing *“Protect your home this monsoon – Get flood cover in Mumbai”* to users in coastal cities.
**Benefit:** Makes the ad more **relevant** and **personalized**, improving click-through and conversion rates.

---

#### **2. Life-Event Based Targeting**

**Meaning:** Show ads to users when a major life event happens (like buying a house, having a baby, or changing jobs).
**Why:** These events often trigger **insurance needs** — for example, new parents may look for life insurance, while new homeowners may seek home coverage.
**Example:** Ads like *“Secure your child’s future with life insurance plans starting ₹499/month.”*
**Benefit:** Helps reach users **at the right moment of need**, improving quote and sales conversion rates.

---

#### **3. Seasonal or Contextual Campaigns**

**Meaning:** Align ad campaigns with **specific seasons or events** that relate to insurance risk or financial activity.
**Why:** During certain times (like **monsoon season**, **tax-filing months**, or **festive periods**), people are more aware of health, safety, or investment needs.
**Example:**

* “Monsoon Health Cover – Stay safe this rainy season.”
* “Tax Saving Life Insurance before March 31st.”
  **Benefit:** Creates **urgency and relevance**, increasing engagement and CTR.

---

#### *(Additional suggestion – optional)*

**4. Retargeting Based on Funnel Stage**
**Meaning:** Serve different ads based on user behavior — people who viewed quotes but didn’t buy can be shown personalized reminders or discount offers.
**Why:** Keeps potential customers engaged and helps **recover drop-offs** in the funnel.

---

### **R – Result:**

These new strategies together can make the campaign more **personalized**, **timely**, and **context-aware**, which helps:

* Improve **click-through rate (CTR)** and **quote generation**
* Reduce **CPA (Cost per Acquisition)** through better targeting
* Increase **brand recall** in key audience groups

---

### **Insight:**

In display advertising, **strategy** simply means *“how you choose whom to show the ad to and what kind of message you show.”*
These new strategies move beyond generic audience targeting to **data-driven personalization**, ensuring that ads are seen by people who are most likely to need the product *at the right place and right time.*

---

Would you like me to write similar **STAR + insight** answers for **7c** (optimisation suggestions) next?


Here’s a **STAR-based and insight-driven answer** for **Question 7c** 👇

---

### **Question 7c:**

*Using the data at your disposal within the data table, suggest more ways to optimise the strategies provided.*

---

### **S – Situation:**

From the campaign dataset, I had access to performance metrics like **Spend, Quotes, Sales, CPA**, and **Strategy names** under two key campaign types — **Prospecting** and **Retargeting**.
The goal was to identify **how to optimise ad spend** and improve **conversion efficiency** using these available metrics.

---

### **T – Task:**

To find **data-backed optimisation actions** that could help allocate budgets effectively, improve conversions, and reduce overall CPA (Cost per Acquisition).

---

### **A – Action:**

Based on the performance data, I suggested the following **optimisation strategies:**

#### **1. Shift Budget Toward High-Performing Strategies**

**What:** Identify and prioritize strategies with **low CPA and high total sales**.
**How:** Use CPA as the key efficiency metric — lower CPA means more cost-effective sales. Reallocate more budget from weaker to stronger performing strategies.
**Why:** Ensures the marketing budget is invested in areas generating the **maximum return on spend**.
**Example:** If *“Audience – Mums”* shows high sales with a lower CPA than *“Homepage 1 Day Recency”*, move more funds there.

---

#### **2. Evaluate and Rebalance Campaign Types (Prospecting vs. Retargeting)**

**What:** Compare the efficiency between **Prospecting** and **Retargeting** campaigns.
**How:** Assess which delivers **cheaper sales and higher conversion rate**. If Retargeting shows lower CPA, slightly increase its share of the total budget.
**Why:** This helps **maximize efficiency** at the funnel’s bottom stage while ensuring Prospecting continues to bring in new audiences.
**Example:** Retargeting CPA = ₹120 vs. Prospecting CPA = ₹150 → allocate more budget to Retargeting.

---

#### **3. Test and Refine Audience Segments**

**What:** Analyse which audience segments (e.g., *Cautious Homeowners*, *Mums*, *Insurance-Keyword*) generate **better quotes and sales volume**.
**How:** Use quote-to-sale ratio and CPA per segment as indicators.
**Why:** Optimising audience selection ensures targeting the **most profitable user groups**, improving both reach and ROI.
**Example:** If *“Cautious Homeowners”* converts 30% more from quote to sale than *“In-Market Insurance”*, prioritize that segment in future campaigns.

---

### **R – Result:**

By implementing these optimisations:

* Budgets can be **reallocated more strategically**, leading to higher ROI.
* **Sales efficiency** can improve by focusing on the best-performing strategies.
* **CPA** can be reduced by removing underperforming segments and focusing on proven ones.

---

### **Insight:**

These optimisations show a shift from a **broad campaign strategy** to a **data-driven allocation model**.

* Instead of evenly distributing spend, we **let performance metrics drive decisions**.
* Comparing Prospecting vs. Retargeting helps balance **funnel growth and conversion efficiency**.
* Audience testing keeps the strategy **dynamic and adaptive** to changing behavior patterns.

---

Would you like me to create the same **STAR + insight answer** for **7d** (“What other campaign data could you look at to make optimisations from?”)?


Here’s a **STAR-based and insight-driven answer** for **Question 7d** 👇

---

### **Question 7d:**

*We have only included a few metrics for this campaign in the “Data” tab (Strategy, Seller etc.).
What other campaign data could you look at to make optimisations from?*

---

### **S – Situation:**

In the dataset provided, only limited metrics such as **Strategy**, **Spend**, **Quotes**, **Sales**, and **CPA** were available.
While these metrics helped evaluate cost efficiency, they didn’t provide enough context on **who** was engaging with the ads, **where** they were engaging from, or **when** conversions were happening.

---

### **T – Task:**

To identify **additional data points** that would enable **deeper campaign optimisation** — helping improve audience targeting, timing, and regional strategy alignment.

---

### **A – Action:**

I identified three key categories of data that would significantly enhance the optimisation process:

#### **1. Demographic Data**

**What:** Include variables like **age, gender, income level, household type**, or **occupation**.
**Why:** This helps us understand which audience group responds best to our campaign.
**How it helps:** We can personalise creatives and messaging — for example, insurance plans for *young parents* vs. *retired homeowners*.
**Example:** If women aged 30–45 are showing higher quote-to-sale conversion, more budget can be allocated to similar profiles.

---

#### **2. Geographic Data**

**What:** Include **region, country, city**, or even **zip code-level targeting**.
**Why:** Product demand and conversion likelihood can vary by location due to local needs or weather conditions.
**How it helps:** Enables **location-based targeting** and **regional ad creatives**.
**Example:** Showing *home insurance* ads to flood-prone areas or *vehicle insurance* ads in urban regions with high traffic density.

---

#### **3. Time-of-Day / Day-of-Week Data**

**What:** Capture **timestamp of clicks or impressions**.
**Why:** Allows understanding of when audiences are most active or responsive.
**How it helps:** Enables **time-based bid optimisation**, ensuring higher ad spend during high-engagement periods.
**Example:** If most conversions occur between 6–9 PM, focus ad delivery during that window to maximise ROI.

---

### **R – Result:**

By integrating demographic, geographic, and time-based data, campaign managers can:

* Achieve **higher targeting precision** and **lower CPA**.
* Personalise ads for **specific audience segments**.
* Increase **conversion rates** by focusing on the right people, in the right place, at the right time.

---

### **Insight:**

This approach transforms the campaign from being **strategy-driven** to **audience-intelligent**.

* Instead of relying only on spend and CPA metrics, we begin to understand **why** certain strategies perform better.
* These extra data layers reveal **hidden drivers of performance** like audience behaviour patterns and geographic trends.
* Ultimately, this allows for **predictive optimisation** — anticipating where and when to spend before inefficiencies arise.

---

Would you like me to write a concise **1-paragraph version** of this answer suitable for direct Excel or report submission (while keeping the STAR essence)?
